specs/makeStyles.md

```diff
@@ -0,0 +1,290 @@
+4. `useCSS` finds the style object in the dictionary, merges the original style object with the child's style overrides and returns the new classname. Instead of performing an expensive object deep merge, it just concatenates the two stringified style objects and depends on the-latest-wins CSS rule.
+
+![](makeStyles-useCss.png)
+
+#### Limitations
+
+This approach looks good for style overrides, we are not sure this approach works the best for core component styles. Without that, we would have two different styling approaches in place.
+Although the object string concatenation is performant, it results in a lot of CSS code.
+
+### v8 `makeStyles`
+
+In order to improve rendering performance, v8 approach is closer to static CSS.
+
+There are two main functions in this approach `makeStyles` and `useStyles`.
+`makeStyles` is a factory which takes style object as input and returns a `useStyles` hook. `makeStyles` does not accept any theme or React context and should be called in a module scope, not in a scope of a React component.
+When rendering a component, `useStyles` hook is called passing a theme to it. The `useStyles` function injects the styles into DOM (if not done yet) and maps props to classnames.
+
+#### Multiple classes are applied to a single element
+
+As in static CSS approach, multiple semantic classes are applied to an element.
+For example `<Button primary disabled>` renders as `<button class="primary disabled">`.
+To style the component, combined classes are used in CSS as well:
+
+```css
+.primary {
+}
+.disabled {
+}
+.primary.disabled {
+}
+```
+
+This is causing problems with overrides - assuming it is guaranteed that override classes are written bellow the component styles in the DOM, for most cases in this example, it is enough to have 1 class specificity for overrides to work correctly. The only case where the 1 class override does not work is for CSS properties in `.disabled.button`.
+Consumer does not know what specificity must be used for the overrides, this can also change between library versions.
+
+#### Stylesheet is generated build time
+
+There is a build time step which generates stylesheet.
+In runtime only the classes are applied based on props and state.
+
+#### Styles precedence is set by module evaluation order
+
+The order of `makeStyles` calls defines the DOM insertion order. When `makeStyles` is called, it does NOT insert the CSS to DOM, but assigns a counter-based id to the associated styles. When `useStyles` returned by the `makeStyles` is called for the first time, it inserts the CSS to DOM in the order defined by the `makeStyles` call.
+
+For that to work as expected, you cannot export `makeStyles` to share styles in multiple different modules - that would lead to non-deterministic insertion order. The only way to share styles is to export the style object and have a dedicated `makeStyles` calls in all the modules which reuse the style object.
```

Comment:
Should we also add notes about async API (React.Suspense)?

https://codesandbox.io/s/kind-davinci-xvwxg